### DISCUSSION TITLE
power-of-two flag in texture compression cli commands

### DISCUSSION THREAD

**Discussion by LoJunKai** [donmccurdy/glTF-Transform#364] [General]
**Description**
Hi, I would just like to clarify the `--power-of-two` flag in the texture compression cli commands. It is stated in the docs that it: 
>Resizes any non-power-of-two textures to the closest power-of-two dimensions, **not exceeding 2048x2048px**. Required for  compatibility on some older devices and APIs, particularly WebGL 1.0.

I was just wondering why is there a 2048x2048px restriction for this closest power-of-two resizing.

**To Reproduce**
1. Model used: https://drive.google.com/file/d/1gULSZ7VRz06Aq9-7oG7T3soDCCN7GxuU/view?usp=sharing
2. Run the following command: `gltf-transform etc1s ./MosquitoInAmber/MosquitoInAmber.gltf  ./MosquitoInAmber-resize/MosquitoInAmber.gltf --power-of-two -v`
2.  Observe that `MosquitoInAmber0.jpg` is being resized by `--resize 4096x2048` in the debug messages

*model taken from: https://sketchfab.com/3d-models/mosquito-in-amber-53a61d58c09b4d2ab30e269aa3e22078
*manually resized `MosquitoInAmber0.jpg` to 4100x4000px

**Question**
In the [source code](https://github.com/donmccurdy/glTF-Transform/blob/80c24086a757e109b780a9fe52cd656ac6e0f50e/packages/cli/src/transforms/toktx.ts#L322) it is written that
```
    if (hi > max) return lo;
    if (hi - value > value - lo) return lo;
    return hi;
```

Why is the first line needed? Wouldn't it be better to return the closest power-of-two all regardless of input dimension?

Thank you for your time! :)

**Comment by donmccurdy** [comment]
Hm, `4100x4000` → `4096x2048` is admittedly an odd-feeling result here. Some background on the thing I'm trying to avoid: My general impression is that...

1. Textures >4K do not render correctly on a pretty significant number of mobile devices.
2. Textures >2K do not render correctly on a small but non-trivial number of mobile devices.

That being the case, I didn't really want to upscale an image silently in a way that's going to reduce device support. At the time I added KTX2/Basis support, this library didn't have any _other_ way to resize images... it does now, so maybe this could be addressed with a simple warning about larger images, advising that the user resize them first if needed. I'm assuming `4096x4096` is the result you would want in this case?

> **EDIT:** Fixed in https://github.com/donmccurdy/glTF-Transform/commit/c1c66fbbaac09c115ce61cd5ed43bcef05677d61, this now warns rather than limiting to 4K.

**Comment by LoJunKai** [comment]
Hey Don, really appreciate the explanation! Yep, I was looking for it to be resized to `4096x4096` instead, the warning would indeed be helpful!

Thank you and have a great day :)

**Reply by donmccurdy** [reply]
Sounds good! This should be fixed shortly by https://github.com/donmccurdy/glTF-Transform/commit/c1c66fbbaac09c115ce61cd5ed43bcef05677d61.
